### DISCUSSION TITLE
Milvus有连接池的概念吗？有的话可以怎么设置呢

### DISCUSSION THREAD

**Discussion by zyy20191** [milvus-io/milvus#18034] [Q&A and General discussion] [answered]
如题，一般用PyMilvus操作，请问可以怎么设置连接池

**Comment by yhmo** [comment, marked as answer]
1.x版本的python sdk我们有实现了简易的连接池，在2.0版本的sdk里没有提供。
不过你可以在2.0版本的ptthon sdk里，指定使用不同的连接来进行远程rpc调用，通过连接的别名(alias)来指定。比如下面这段，在构造Collection对象的时候使用了一个参数"using"，这个参数就是用来指定一个连接的别名，这个collection对象所有的rpc调用都使用这个连接：
```
from pymilvus import (
    connections,
    Collection,
}

connections.connect(host=_HOST, port=_PORT, using="my_connection")

collection = Collection(name=name, schema=schema, using="my_connection")
collection.load()
```

connections.connect(host=_HOST, port=_PORT, using="my_connection") 就是创建一个连接，其别名为"my_connection"
connections里面有一个map，别名为key，连接为value。通过这种方式，创建多个连接。你可以让每个Collection对象使用一个单独的连接，也可以共享同一个连接。也可以给你的每个工作线程里创建并使用一个连接。

**Reply by zyy20191** [reply]
那如果连接不关闭的话，请问有自动断开连接的设置吗，大概多久会自动断开连接呢

**Reply by yhmo** [reply]
在pymilvus项目代码里，pymilvus/client/grpc_handler.py 这个文件是用来建立grpc连接的：

```
class GrpcHandler:
        def _setup_grpc_channel(self):
        """ Create a ddl grpc channel """
        if self._channel is None:
            opts = [(cygrpc.ChannelArgKey.max_send_message_length, -1),
                    (cygrpc.ChannelArgKey.max_receive_message_length, -1),
                    ('grpc.enable_retries', 1),
                    ('grpc.keepalive_time_ms', 55000),
                    ]
```

可以看到，这里设置的keepalive_time_ms=55000ms，也就是55秒。

**Reply by zyy20191** [reply]
好的谢谢！

**Reply by yhmo** [reply]
等等，我刚查了一下，这个设置有点问题。
从grpc的文档 https://github.com/grpc/grpc/blob/master/doc/keepalive.md 上看：
- keepalive_time_ms是定时发ping保持连接的间隔时间，客户端默认是无限大
- keepalive_timeout_ms才是关闭连接的时间，客户端默认是20秒

所以这里设置'grpc.keepalive_time_ms'=55000 实际无效，连接是仍然20秒后关闭。之前写这个代码的人可能错以为keepalive_time_ms是保持连接的时间长短了。后面这里改掉。

**Reply by zyy20191** [reply]
ohhhh！谢谢，费心了
